File: projects/01-smart-agriculture/fog/aggregation.py
from statistics import fmean


def aggregate(sensor_type, site_id, unit, readings, window_start, window_end):
    """Reduce one window's buffered readings (a plain list of {"ts", "value"}
    dicts accumulated by app.py's ingest handler) into one summary record.

    latest is the last reading in arrival order, not the max-timestamp
    reading -- sensors dispatch batches in order, so arrival order already
    reflects chronological order without parsing/sorting the ts strings.
    """
    values = [r["value"] for r in readings]
    return {
        "sensor_type": sensor_type,
        "site_id": site_id,
        "unit": unit,
        "window_start": window_start,
        "window_end": window_end,
        "count": len(values),
        "min": min(values),
        "max": max(values),
        "avg": round(fmean(values), 3),
        "latest": readings[-1]["value"],
    }

File: projects/10-wildfire-forest-monitoring/fog/aggregation.js
"use strict";

// Pure window-aggregate math, kept separate from the buffering mechanism
// (buffer.js) so it can be unit tested against plain arrays of values
// without touching the EventEmitter/Map plumbing at all.
function summarizeWindow(sensorType, siteId, unit, readings, windowStart, windowEnd) {
  const values = readings.map((r) => r.value);
  const sum = values.reduce((acc, v) => acc + v, 0);
  return {
    sensor_type: sensorType,
    site_id: siteId,
    unit,
    window_start: windowStart,
    window_end: windowEnd,
    count: values.length,
    min: Math.min(...values),
    max: Math.max(...values),
    avg: Math.round((sum / values.length) * 1000) / 1000,
    latest: readings[readings.length - 1].value,
  };
}

module.exports = { summarizeWindow };
